### DISCUSSION TITLE
Help with spacy-llm Llama2 config

### DISCUSSION THREAD

**Discussion by fireblonde** [explosion/spacy-llm#324] [Q&A]
Hi, 

I am following the tutorial for Llama2. My config is as follows:

```python3
[nlp]
lang = "en"
pipeline = ["llm"]
batch_size = 15

[components]

[components.llm]
factory = "llm"

[components.llm.task]
@llm_tasks = "spacy.TextCat.v2"
labels = ["INSULT", "THREAT"]

[components.llm.model]
@llm_models = "spacy.Llama2.v1"
name = "Llama-2-13b-hf"
```

and I am getting this error:

```bash
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 136.00 MiB (GPU 0; 23.67 GiB total capacity; 22.21 GiB already allocated; 157.81 MiB free; 22.22 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

I tried using all of the available llama2 models. If I use the smallest model, then I get this error (similar to another discussion):
```bash
ValueError: The following `model_kwargs` are not used by the model: ['use_auth_token'] (note: typos in the generate arguments will also show up in this list)
```

Can you please help me with this? Is there a possibility to use `accelerate` and if yes, can you help me with a more detailed instructions? 

Thank you so much in advance! Looking forward to your answers! :)

**Comment by rmitsch** [comment]
Hi @fireblonde! Let's start with the second error message. Have you 
- requested and been granted permission to use the Llama 2 model and
- are logged in to your Hugging Face account on the machine you're running this from?

See the "Gated models on Hugging Face" section [here](https://spacy.io/api/large-language-models#models-hf).

**Reply by fireblonde** [reply]
Hi @rmitsch,

Thank you for the prompt answer!
The answers to both questions are yes, 

- I do login via the cli, also tried from python code.

- I do have the three models (non-chat versions) 

Thanks :)

**Reply by fireblonde** [reply]
Also @rmitsch, 
Following this discussion (Error using llama2 #273), the code where the tokeniser is provided with the Breaking Bad example is running for me in 7b and  13b without running out of memory or setting batch size (70b does but if the settings are configured it works)

**Reply by rmitsch** [reply]
Thanks for the info. This seems indeed similar to #273, but so far I couldn't reproduce this problem.
Can you post which versions of `spacy-llm` and `transformers` you're using?

**Reply by fireblonde** [reply]
Yes, of course!

I have `spacy-llm` Version `0.5.1` and 
`transformers` Version `4.33.2`. 

Thank you!

**Reply by rmitsch** [reply]
Do you have the same issue after upgrading to `spacy-llm` 0.6.0?

**Reply by fireblonde** [reply]
I do get them: 
- with 13b I get `Out of memory`,
- with the 7b, I get the `ValueError`

**Reply by rmitsch** [reply]
@fireblonde  [v0.6.1](https://github.com/explosion/spacy-llm/releases/tag/v0.6.1) is out. Give it a try and let me know if that fixes the `ValueError`?

**Reply by fireblonde** [reply]
The `ValueError` seems to not appear if I do the 7b model now with the `spacy-llm` v0.6.1- I can see a proper output now :).
 However, with 13b and 70b it persists with 
```bash
the torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 136.00 MiB (GPU 0; 23.67 GiB total capacity; 22.21 GiB already allocated; 143.12 MiB free; 22.22 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

Do you have any suggestions for that - anything to set in the .cfg file or more information on how to run it on cpu instead? 

Thanks in advance!

**Reply by rmitsch** [reply]
Ok, that's progress :slightly_smiling_face: 

Regarding the OOM error - `spacy-llm` wraps Hugging Face models. Everything you pass to the model in `config_init` is passed on to the HF model at init time. The same applies to `config_run` at run time.

The Llama 2 model is loaded and executed via [`transformers.pipeline()`](https://huggingface.co/transformers/v3.0.2/main_classes/pipelines.html). It accepts a `device` parameter with the device ID the model should run on, with -1 representing your CPU. So this should make your model run on your CPU:
```ini
[components.llm.model]
@llm_models = "spacy.Llama2.v1"
name = "Llama-2-13b-hf"
config_init = {"device": -1}
```
I haven't run Llama 2 on a CPU yet. You should expect the model to respond very slowly.

**Reply by fireblonde** [reply]
Some progress - 13b and 7b works with CPU! Indeed, it is very slow but works! Thank you for the solution!

I am still unsure on how to solve the GPU out of memory issue and make the 70b work :(

I tried to follow this structure which works outside of using `spacy-llm`:
```python
tokenizer = AutoTokenizer.from_pretrained(model)
pipeline = transformers.pipeline(
    "text-generation",
    model=model,
    torch_dtype=torch.float16,
    device_map="auto",
)
```
but no luck so far.

**Reply by rmitsch** [reply]
Have you monitored how much memory is allocated in your system when you load the 7b and 13b versions? ~Your system possibly doesn't have enough RAM to hold the 70b model.~

Sorry, I think I misunderstood. You _can_ run the 70b model outside of `spacy-llm`?

**Reply by fireblonde** [reply]
Yes, I can run 70b outside of `spacy-llm` with around 20922MiB / 24564MiB Memory Usage.

**Reply by rmitsch** [reply]
Alright. Have you tried passing the `device_map` setting to `spacy_llm`? I. e. 
```ini
...
config_init = {"device_map": "auto"}
...
```

Passing `torch.dtype` isn't possible yet, but on our roadmap.

**Reply by fireblonde** [reply]
If I have both the `device` to be `-1` and  `device_map` to be `auto` I get this when running 13b:
```bash
ValueError: The model has been loaded with `accelerate` and therefore cannot be moved to a specific device. Please discard the `device` argument when creating your pipeline object.
```

If I only run it with `device_map` to be `auto` then it outputs the same:
```bash
ValueError: The model has been loaded with `accelerate` and therefore cannot be moved to a specific device. Please discard the `device` argument when creating your pipeline object.
```

**Reply by fireblonde** [reply]
I was also reading on HF website about the models' GPU specs:

They mention:
```
    For 7B models, we advise you to select "GPU [medium] - 1x Nvidia A10G".
    For 13B models, we advise you to select "GPU [xlarge] - 1x Nvidia A100".
    For 70B models, we advise you to select "GPU [2xlarge] - 2x Nvidia A100" with bitsandbytes quantization enabled or "GPU [4xlarge] - 4x Nvidia A100"
```
I work on NVIDIA RTX A5000 :(

**Reply by rmitsch** [reply]
> If I only run it with device_map to be auto then it outputs the same:

That's because we move this model to `cuda:0` by default...I guess that's not a good idea after all since `device_map="auto"` is an option. We'll change that in the next patch version.

**Reply by rmitsch** [reply]
[`spacy-llm` v0.6.3](https://github.com/explosion/spacy-llm/releases/tag/v0.6.3) is out and should fix these device binding issue. I recommend updating the library and giving it another try.
